### ISSUE TITLE
Variant `NC_000010.10:g.11330462A>G` throws an unknown error.

### ISSUE THREAD

**Issue by ifokkema** [openvar/variantValidator#464]
Variant `NC_000010.10:g.11330462A>G` throws an unknown error. When submitting it through the web interface, VV reports:

> Unable to validate the submitted variant NC_000010.10:g.11330462A>G against the GRCh37 assembly.
> Please check your submission and re-submit.

When submitting it through the [VV API](https://rest.variantvalidator.org/VariantValidator/variantvalidator/hg19/NC_000010.10%3Ag.11330462A%3EG/all?content-type=application%2Fjson), the output is:
```json
{
  "message": "Internal Server Error"
}
```

When using the [LOVD endpoint](https://rest.variantvalidator.org/LOVD/lovd/hg19/NC_000010.10%3Ag.11330462A%3EG/refseq/all/true/false?content-type=application%2Fjson), all options turned off, we do get output:

```json
{
  "NC_000010.10:g.11330462A>G": {
    "NC_000010.10:g.11330462A>G": {
      "g_hgvs": "NC_000010.10:g.11330462A>G",
      "genomic_variant_error": null,
      "hgvs_t_and_p": null,
      "p_vcf": "10:11330462:A:G",
      "selected_build": "hg19"
    },
    "errors": [],
    "flag": null
  },
  "metadata": {
    "variantformatter_version": "2.2.0",
    "variantvalidator_hgvs_version": "2.2.0",
    "variantvalidator_version": "2.2.0",
    "vvdb_version": "vvdb_2022_11",
    "vvseqrepo_db": "VV_SR_2022_11/master",
    "vvta_version": "vvta_2022_11"
  }
}
```

When I request mappings, I start getting a few warnings (filtered JSON).

```json
{
  "NC_000010.10:g.11330462A>G": {
    "NC_000010.10:g.11330462A>G": {
      "g_hgvs": "NC_000010.10:g.11330462A>G",
      "genomic_variant_error": null,
      "hgvs_t_and_p": {
        "NM_001025076.2": {
          "gap_statement": null,
          "gapped_alignment_warning": null,
          "p_hgvs_slc": null,
          "p_hgvs_tlc": null,
          "t_hgvs": null,
          "transcript_variant_error": "No alignment data available for transcript NM_001025076.2 and chromosome NC_000010.10"
        },
        "NM_001083591.1": {
          "gap_statement": null,
          "gapped_alignment_warning": null,
          "p_hgvs_slc": null,
          "p_hgvs_tlc": null,
          "t_hgvs": null,
          "transcript_variant_error": "No alignment data available for transcript NM_001083591.1 and chromosome NC_000010.10"
        },
        "NM_006561.3": {
          "gap_statement": null,
          "gapped_alignment_warning": null,
          "p_hgvs_slc": null,
          "p_hgvs_tlc": null,
          "t_hgvs": null,
          "transcript_variant_error": "No alignment data available for transcript NM_006561.3 and chromosome NC_000010.10"
        },
      },
      "p_vcf": "10:11330462:A:G",
      "selected_build": "hg19"
    },
    "errors": [],
    "flag": "processing_error"
  },
  "metadata": {
    "variantformatter_version": "2.2.0",
    "variantvalidator_hgvs_version": "2.2.0",
    "variantvalidator_version": "2.2.0",
    "vvdb_version": "vvdb_2022_11",
    "vvseqrepo_db": "VV_SR_2022_11/master",
    "vvta_version": "vvta_2022_11"
  }
}
```

Adding liftover then kills the output. It's not even JSON, so maybe an webserver timeout.

```json
An internal server error occurred. Please try again later.
```

Although, selecting a single transcript seems to work, even when it's one of those transcripts that threw an error.

**Comment by Peter-J-Freeman** [comment]
I can confirm that for VV the following is throwing an error

```python
import json
import VariantValidator
vval = VariantValidator.Validator()
variant = 'NC_000010.10:g.11330462A>G' # variant 1
genome_build = 'GRCh37'
select_transcripts = 'all'
transcript_set = 'refseq'
validate = vval.validate(variant, genome_build, select_transcripts, transcript_set)
validation = validate.format_as_dict(with_meta=True)
print(json.dumps(validation, sort_keys=True, indent=4, separators=(',', ': '))
```

Falls over completely

**Comment by John-F-Wagstaff** [comment]
Hello @Peter-J-Freeman, just from a quick check on the database end I can confirm that all of these sequences are present in the VVTA database, and available as current data. The alignments for NM_001083591.1 and NM_006561.3 where not touched in the latest database update.

I hope this helps you narrow things down.

**Comment by Peter-J-Freeman** [comment]
Added the following code around here

https://github.com/openvar/variantValidator/blob/master/VariantValidator/modules/gapped_mapping.py#L394

```python
            # Get orientation of the gene wrt genome and a list of exons mapped to the genome


            print(saved_hgvs_coding.ac)

            ori = self.validator.tx_exons(tx_ac=saved_hgvs_coding.ac, alt_ac=self.hgvs_genomic_5pr.ac,
                                          alt_aln_method=self.validator.alt_aln_method)

            print(ori)
```

Get the print out
NM_001025076.2
hgvs Exception: No tx_exons for (tx_ac=NM_001025076.2,alt_ac=NC_000010.10,alt_aln_method=splign)

I will make changes to the code to fix this error. @John-F-Wagstaff, can you look at VVTA for this alignment. Is there anything to indicate that this is a difficult transcript? I will pick this up tomrrow

**Comment by John-F-Wagstaff** [comment]
It looks like something is going wrong with the archival of unused/superseded transcript_exon_set data. The within transcript exon set data is masked, as if superseded, when new data was loaded for one of the genome versions, but is still being used for the older genome, but only very rarely.  The mapping is left un-archived, when if it is superseded it should be masked too. 

It should be possible to, relatively trivially, just unmask all of this problematic excess marked data but I want to know why it happens too, before I count this as fixed.  transcript_exon_set data should only be masked when the last mapping taken from it is masked, not before.

currently this affects 157 transcript mappings (of 708725)  so it is perhaps unsurprising this slipped through our current tests.

**Comment by John-F-Wagstaff** [comment]
OK, so I now know what happened, and why.

TLDR: this is a 1 off problem caused by the original data import, I will get a fixed version of the database done shortly, rebuild the views, and get a new version to @Peter-J-Freeman asap

Details -
The UTA from which, our older data was loaded can only have one idea of the within transcript exon positions at a time, when a new version is loaded the old one must be marked as archived/superseded. This causes problems for the alignment data, particularly when there are GRCh37->GRCh38 differences.  For this reason we had to restructure the database to fix the underlying limitation,  and clean up the data on import. Old versions of the aligned exon position data are also tagged as superseded when new alignments to the same transcript->ref pair are loaded with the same alignment type (e.g. spline).

As a result of our changes the VVTA can have multiple versions of the within transcript exon positions, each mapping is tied to one specific version of the exon potions that it comes from. Mappings can still be superseded by new data as before. Originally I had code in place to mask unused/superseded transcript_exon_set data, when the last mapping for said within transcript exon set was itself superseded. This however added unnecessary complexity, and since we rely on the up to date mappings for the relevant views was redundant.  As such within transcript exon set masking, along with the unmasking for re-used data, was removed, possibly before I finalised the patch. The field was left in the database however for import, and later expansion, purposes.

The intent was that this would not cause any problems.  The old versions of the within transcript exon data could not be linked reliably to it's mappings once archived, and we purged all unused exon data along with the broken alignments during the initial load. This should have left us without any archived within transcript exon data, however in order to not end up removing excess data or losing information we had to load 1 new set of transcript data before doing the purge. This data when it was loaded, like always, used the existing within transcript exon data, when there was a transcript of the same accession and version number with a exon set that matched the current exon data.  66 times this included a exon set that had been archived in the UTA workflow. These 66 within transcript exon position sets (mostly from GRCh37 data) where therefore used by valid alignments, and so escaped the final stage of the later purging, and at this point have 157 mappings between themselves.

With this as the origin of the issue I can just purge the "archived" tags and be sure that 1, the data is valid as is, and 2, the problem won't re-occur. I will get a purged rebuilt version of the database done and posted to the data server once it has been tested.

**Comment by Peter-J-Freeman** [comment]
Makes sense @John-F-Wagstaff . So we will put mark this up as a VVTA issue and I will not apply code changed until we test the next VVTA release?

**Comment by John-F-Wagstaff** [comment]
@Peter-J-Freeman Given the small number of transcripts affected I was just going to unmask them, and suggest a point release.  Since I think you used the transcript list not the mapped transcript or valid transcript data this would not need any changes to the variantValdiatior database to just work. With this done any variantValdiatior code changes would be unnecessary, unless/until we decide to use this marking in normal data.  If you want to push the code changes to fix this for potential future proofing, then testing them before any database updates would be the way to go.

However an extra VVTA database release does mean more testing, we would not want to release without, and keeping/deploying an out of standard extra database release for VVTA. So when you say the next VVTA release do you mean a small patch release for this issue, or do you mean to leave this until our next scheduled major update?

**Comment by i3hsInnovation** [comment]
point release is fine by me. Lets do that. Sorry, logged in as different user at the moment. 

So Do point release ASAP. Do next release of VVTA in ~May? Think that's a 6 month cycle???

**Comment by John-F-Wagstaff** [comment]
I will get on that now, I need to make sure that I am using clean versions of the databases though, so I have a bit of download/reload etc. work.

The timing works fine for me, 6 months was the plan from the start, as long as you had enough people to do it.

**Comment by John-F-Wagstaff** [comment]
@Peter-J-Freeman  I have created, tested, and uploaded to the lamp a point release of VVTA to fix this issue . It tests OK with the standard variantValidatior tests, and produces good results for the test code for this bug that you provide above. The database file name is vvta_2022_11.1_no_seq.sql.gz and the schema name is vvta_2022_11_1.  If this works for your tests as well then it should fix this on our main server too, when uploaded to it.

**Comment by Peter-J-Freeman** [comment]
I will test on my system. Once it all pans out I will make necessary releases on the servers

**Comment by Peter-J-Freeman** [comment]
@ifokkema . On server ready for testing

**Comment by ifokkema** [comment]
I observed no change - the site and REST API behaves the same as when the issue was reported (errors and warnings).

**Comment by Peter-J-Freeman** [comment]
I think that will be attributable to human error. Have a feeling I forgot to tell Apache about the new database build. On it

**Comment by Peter-J-Freeman** [comment]
OK try this

https://rest.variantvalidator.org/VariantValidator/variantvalidator/GRCh38/NC_000010.10%3Ag.11330462A%3EG/NM_001025076.2?content-type=application%2Fjson

You still get a time-out when you try and submit without selecting transcripts, but this gene has >30 transcript variants with several versions, so I'm not surprised

**Comment by ifokkema** [comment]
OK, thanks! So I guess we're going from one error into the next, but at least the underlying issue was solved, and now we run into the general issue of receiving a timeout when too much work needs to be done. So status is now:

- Interface fails after timeout unless transcripts are selected.
- LOVD API endpoint interface works when options are off.
- When trying to map to all known transcripts without protein prediction or liftover, the LOVD endpoint now fails (which wasn't the case before). Probably more processing is done than before, and the timeout is reached.
- The LOVD API endpoint with mapping and transcripts selected works now, `transcript_variant_error` is now `null`. Liftover works, too.

So I think, acceptable? It would be great if, in the future, VV could "detect" when something will take too long and reduce the number of transcripts processed or otherwise return something that would allow the user to obtain a list of transcripts to choose from. If you like this idea, I can submit an issue if you want.

**Comment by i3hsInnovation** [comment]
Also, I suspect that yes, you are now hitting timeouts because there are now lots of valid transcripts that VV could not use before
